t/semaphore.t

```diff
@@ -0,0 +1,1068 @@
+                end,
+                sem,ret)
+            ngx.thread.wait(co1)
+            ngx.thread.wait(co2)
+            if ret.post and ret.wait then
+                ngx.say("right")
+            end
+            ngx.exit(200)
+        ';
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+right
+--- no_error_log
+[error]
+
+
+
+=== TEST 10: basic semaphore wait post in rewrite_by_lua
+--- http_config eval: $::HttpConfig
+--- config
+    location = /test {
+        access_log off;
+        content_by_lua '
+        local res1, res2 = ngx.location.capture_multi{
+          { "/sem_wait"},
+          { "/sem_post"},
+        }
+        ngx.say(res1.status)
+        ngx.say(res1.body)
+        ngx.say(res2.status)
+        ngx.say(res2.body)
+        ';
+    }
+
+    location /sem_wait {
+        rewrite_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            if not g.test then
+                local sem, err = semaphore.new(0)
+                if not sem then
+                    ngx.say(err)
+                    ngx.exit(500)
+                end
+                g.test = sem
+            end
+            local sem = g.test
+            local ok, err = sem:wait(10)
+            if ok then
+                ngx.print("wait")
+                ngx.exit(200)
+            else
+                ngx.print("some badthing happen"..err)
+                ngx.exit(201)
+            end
+        ';
+    }
+
+    location /sem_post {
+        rewrite_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            if not g.test then
+                local sem, err = semaphore.new(0)
+                if not sem then
+                    ngx.say(err)
+                    ngx.exit(500)
+                end
+                g.test = sem
+            end
+            local sem = g.test
+             local ok, err = sem:post()
+            if ok then
+                ngx.print("post")
+                ngx.exit(200)
+            else
+                ngx.print("some badthing happen"..err)
+                ngx.exit(201)
+            end
+    ';
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+200
+wait
+200
+post
+--- no_error_log
+[error]
+
+
+
+=== TEST 11: basic semaphore wait in timer.at
+--- http_config eval: $::HttpConfig
+--- config
+    location /test {
+       content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local sem = semaphore.new(0)
+            local g = getmetatable(_G).__index
+            local ok, err = sem:post()
+            g.test = sem
+            local function func(premature, g)
+                local sem1 = g.test
+                local ok, err = sem1:wait(10)
+                if not ok then
+                    ngx.log(ngx.ERR, err)
+                end
+            end
+            ngx.timer.at(0,func,g)
+            ngx.sleep(2)
```

Comment:
I don't think we have to wait as long as 2 seconds in this test case. The running time of tests is important since the tests are supposed to run over and over again.